<center>
    
# 人工智能高阶人才培训班
## 课程作业（二）

---

> 💥说明💥
> 1. 本次作业最晚提交时间： **<font color='red'>2019年6月12日</font>**
> 2. 作业中以`Q-n`标注的几个问题的答案请通过[答题页面](https://app.askform.cn/43245470-5adc-4293-9427-0a33e9743625.aspx?Type=2)连接的Form表格提交，提交中填写的姓名请与本人姓名保持一致；
> 3. 以`Lesson02-HW-{你的名字英文}.ipynb`命名的作业文件请参考[文档](../How-To.md)说明提交到Git的代码仓库；
> 4. 如果完成作业过程中，你有任何新的想法和解题思路，可以完成一篇分享文档(建议用MarkDown撰写)，将根据质量给予⭐奖励
> 5. 🚫 **注意：** 请不要直接编辑该模板文件，建议复制该模板文件生成新的以你自己姓名命名的作业文件（`Lesson02-HW-{你的名字中文}.ipynb`）

### Q1
参考课程中使用`numpy`进行数理统计的过程，使用numpy提供的库`random.normal`生成1万个样本

本题目中使用下表格中**华中地区的女性身高数据**作为参考源

<center>

##### 中国成年人身高，体重均值和标准差的参考值 （2015）
|性别|东北华北|西北|东南|华中|华南|西南|
|-|-|-|-|-|-|-|
|身高，体重|均值，标准差|均值，标准差|均值，标准差|均值，标准差|均值，标准差|均值，标准差|
|男(mm)|1693, 56.6|1684, 53.7|1686, 55,2|1669, 56.3|1650, 57.1|1647, 56.7|
|女(mm)|1586, 51.8|1575, 51.9|1575, 50.8|1560, 50.7|1549, 49.7|1546, 53.9|
|男(kg)|64, 8.2|60, 7.6|59, 7.7|57, 6.9|56, 6.9|55, 6.8|
|女(kg)|55, 7.1|52, 7.1|51, 7.2|50, 6.8|49, 6.5|50, 6.9|


> **注意：**为保证结果的一致性，请每次在你执行随机分布生成之前，使用`np.random.seed(42)`，产生同样的种子值

#### 问： 你生成的分布的实际产生的方差、均值、中位数各是多少(保留两位小数)？
- 50.70, 1560.01, 1575.91
- 50.87, 1560, 1560.69
- 2588.06, 1559.89, 1559.87
- 50.87, 1559.89, 1559.87

In [10]:
import numpy as np
np.random.seed(42)
# your code
height_10000=np.random.normal(1560, 50.7, 10000) 
print("variance:",round(np.var(height_10000),2))#方差
print("mean:",round(np.mean(height_10000),2))#均值
print("median:",round(np.median(height_10000),2))#中值

variance: 2588.06
mean: 1559.89
median: 1559.87


## Q2-Q9
Q2-Q9将使用kaggle入门中常使用的泰坦尼克数据集，详情可以访问[Titanic: Machine Learning from Disaster](https://www.kaggle.com/c/titanic)，为了方便处理我们已经对原始数据进行了初步的处理（如缺失值、异常值等）。

在这一系列题目中，我们将综合使用`numpy`, `pandas`和`matplotlib`解决一个泰坦尼克号数据的初步分析问题。在机器学习中，通过多种不同的方式对数据进行分析有助于及时捕获数据特征和数据之间的关联。

In [11]:
import numpy as np
import pandas as pd
from matplotlib import pyplot as plt

首先我们使用pandas导入数据

In [14]:
data = pd.read_csv('../data/titanic_train.csv',
                  index_col='PassengerId')

查看数据结构和基本组成

In [15]:
data.head(5)

Survived  Pclass  \
PassengerId                     
1                   0       3   
2                   1       1   
3                   1       3   
4                   1       1   
5                   0       3   

                                                          Name     Sex   Age  \
PassengerId                                                                    
1                                      Braund, Mr. Owen Harris    male  22.0   
2            Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0   
3                                       Heikkinen, Miss. Laina  female  26.0   
4                 Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0   
5                                     Allen, Mr. William Henry    male  35.0   

             SibSp  Parch            Ticket     Fare Cabin Embarked  
PassengerId                                                          
1                1      0         A/5 21171   7.2500   NaN        S  
2                1      0          PC 17599  71.2833   C85        C  
3                0      0  STON/O2. 3101282   7.9250   NaN        S  
4                1      0            113803  53.1000  C123        S  
5                0      0            373450   8.0500   NaN        S

每个字段的含义如下：

| **Variable** | **Definition**                             | **Key**                                        |
| ------------ | ------------------------------------------ | ---------------------------------------------- |
| survival     | Survival                                   | 0 = No, 1 = Yes                                |
| pclass       | Ticket class                               | 1 = 1st, 2 = 2nd, 3 = 3rd                      |
| sex          | Sex                                        |                                                |
| Age          | Age in years                               |                                                |
| sibsp        | # of siblings / spouses aboard the Titanic |                                                |
| parch        | # of parents / children aboard the Titanic |                                                |
| ticket       | Ticket number                              |                                                |
| fare         | Passenger fare                             |                                                |
| cabin        | Cabin number                               |                                                |
| embarked     | Port of Embarkation                        | C = Cherbourg, Q = Queenstown, S = Southampton |


In [17]:
data.describe()

Survived      Pclass         Age       SibSp       Parch        Fare
count  891.000000  891.000000  714.000000  891.000000  891.000000  891.000000
mean     0.383838    2.308642   29.699118    0.523008    0.381594   32.204208
std      0.486592    0.836071   14.526497    1.102743    0.806057   49.693429
min      0.000000    1.000000    0.420000    0.000000    0.000000    0.000000
25%      0.000000    2.000000   20.125000    0.000000    0.000000    7.910400
50%      0.000000    3.000000   28.000000    0.000000    0.000000   14.454200
75%      1.000000    3.000000   38.000000    1.000000    0.000000   31.000000
max      1.000000    3.000000   80.000000    8.000000    6.000000  512.329200

为了让大家快速熟悉pandas，我们准备了几个简单的操作：

1. 让我们尝试从乘客中过滤只显示通过Cherbourg(Embarked=C) 登陆并且付费大于200磅（fare > 200）：

In [19]:
data[(data['Embarked'] == 'C') & (data.Fare > 200)].head()

Survived  Pclass  \
PassengerId                     
119                 0       1   
259                 1       1   
300                 1       1   
312                 1       1   
378                 0       1   

                                                        Name     Sex   Age  \
PassengerId                                                                  
119                                 Baxter, Mr. Quigg Edmond    male  24.0   
259                                         Ward, Miss. Anna  female  35.0   
300          Baxter, Mrs. James (Helene DeLaudeniere Chaput)  female  50.0   
312                               Ryerson, Miss. Emily Borie  female  18.0   
378                                Widener, Mr. Harry Elkins    male  27.0   

             SibSp  Parch    Ticket      Fare            Cabin Embarked  
PassengerId                                                              
119              0      1  PC 17558  247.5208          B58 B60        C  
259              0      0  PC 17755  512.3292              NaN        C  
300              0      1  PC 17558  247.5208          B58 B60        C  
312              2      2  PC 17608  262.3750  B57 B59 B63 B66        C  
378              0      2    113503  211.5000              C82        C

2. 我们可以使用`sort_values`降序排列这些数据，排序关键字通过`by`字段定义

In [22]:
data[(data['Embarked'] == 'C') & 
     (data['Fare'] > 200)].sort_values(by='Fare',
                               ascending=False).head()

Survived  Pclass                                   Name     Sex  \
PassengerId                                                                    
259                 1       1                       Ward, Miss. Anna  female   
680                 1       1     Cardeza, Mr. Thomas Drake Martinez    male   
738                 1       1                 Lesurer, Mr. Gustave J    male   
312                 1       1             Ryerson, Miss. Emily Borie  female   
743                 1       1  Ryerson, Miss. Susan Parker "Suzette"  female   

              Age  SibSp  Parch    Ticket      Fare            Cabin Embarked  
PassengerId                                                                    
259          35.0      0      0  PC 17755  512.3292              NaN        C  
680          36.0      0      1  PC 17755  512.3292      B51 B53 B55        C  
738          35.0      0      0  PC 17755  512.3292             B101        C  
312          18.0      2      2  PC 17608  262.3750  B57 B59 B63 B66        C  
743          21.0      2      2  PC 17608  262.3750  B57 B59 B63 B66        C

3. 我们来尝试在数据中添加一个新的特征（一个新的column）

In [5]:
def age_category(age):
    '''
    < 30 -> 1
    >= 30, <55 -> 2
    >= 55 -> 3
    '''
    if age < 30:
        return 1
    elif age < 55:
        return 2
    elif age >= 55:
        return 3

In [6]:
age_categories = [age_category(age) for age in data.Age]
data['Age_category'] = age_categories

另一种实现方法是通过`.apply`函数实现
```python
data['Age_category'] = data['Age'].apply(age_category)
```

你可以通过`data.head()`命令查看新的数据的情况

In [7]:
data.head()

Survived  Pclass  \
PassengerId                     
1                   0       3   
2                   1       1   
3                   1       3   
4                   1       1   
5                   0       3   

                                                          Name     Sex   Age  \
PassengerId                                                                    
1                                      Braund, Mr. Owen Harris    male  22.0   
2            Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0   
3                                       Heikkinen, Miss. Laina  female  26.0   
4                 Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0   
5                                     Allen, Mr. William Henry    male  35.0   

             SibSp  Parch            Ticket     Fare Cabin Embarked  \
PassengerId                                                           
1                1      0         A/5 21171   7.2500   NaN        S   
2                1      0          PC 17599  71.2833   C85        C   
3                0      0  STON/O2. 3101282   7.9250   NaN        S   
4                1      0            113803  53.1000  C123        S   
5                0      0            373450   8.0500   NaN        S   

             Age_category  
PassengerId                
1                     1.0  
2                     2.0  
3                     1.0  
4                     2.0  
5                     2.0

### **Q2. 船上有多少男性和女性？**
- 412 男性 479 女性
- 314 男性 577 女性
- 479 男性 412 女性
- 577 男性 314 女性

In [30]:
# 在此编写代码回答问题
#
data[data["Sex"]=="male"]
data[data["Sex"]=="female"]

Survived  Pclass  \
PassengerId                     
2                   1       1   
3                   1       3   
4                   1       1   
9                   1       3   
10                  1       2   
11                  1       3   
12                  1       1   
15                  0       3   
16                  1       2   
19                  0       3   
20                  1       3   
23                  1       3   
25                  0       3   
26                  1       3   
29                  1       3   
32                  1       1   
33                  1       3   
39                  0       3   
40                  1       3   
41                  0       3   
42                  0       2   
44                  1       2   
45                  1       3   
48                  1       3   
50                  0       3   
53                  1       1   
54                  1       2   
57                  1       2   
59                  1       2   
62                  1       1   
...               ...     ...   
808                 0       3   
810                 1       1   
814                 0       3   
817                 0       3   
821                 1       1   
824                 1       3   
830                 1       1   
831                 1       3   
836                 1       1   
843                 1       1   
850                 1       1   
853                 0       3   
854                 1       1   
855                 0       2   
856                 1       3   
857                 1       1   
859                 1       3   
863                 1       1   
864                 0       3   
866                 1       2   
867                 1       2   
872                 1       1   
875                 1       2   
876                 1       3   
880                 1       1   
881                 1       2   
883                 0       3   
886                 0       3   
888                 1       1   
889                 0       3   

                                                          Name     Sex   Age  \
PassengerId                                                                    
2            Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0   
3                                       Heikkinen, Miss. Laina  female  26.0   
4                 Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0   
9            Johnson, Mrs. Oscar W (Elisabeth Vilhelmina Berg)  female  27.0   
10                         Nasser, Mrs. Nicholas (Adele Achem)  female  14.0   
11                             Sandstrom, Miss. Marguerite Rut  female   4.0   
12                                    Bonnell, Miss. Elizabeth  female  58.0   
15                        Vestrom, Miss. Hulda Amanda Adolfina  female  14.0   
16                            Hewlett, Mrs. (Mary D Kingcome)   female  55.0   
19           Vander Planke, Mrs. Julius (Emelia Maria Vande...  female  31.0   
20                                     Masselmani, Mrs. Fatima  female   NaN   
23                                 McGowan, Miss. Anna "Annie"  female  15.0   
25                               Palsson, Miss. Torborg Danira  female   8.0   
26           Asplund, Mrs. Carl Oscar (Selma Augusta Emilia...  female  38.0   
29                               O'Dwyer, Miss. Ellen "Nellie"  female   NaN   
32              Spencer, Mrs. William Augustus (Marie Eugenie)  female   NaN   
33                                    Glynn, Miss. Mary Agatha  female   NaN   
39                          Vander Planke, Miss. Augusta Maria  female  18.0   
40                                 Nicola-Yarred, Miss. Jamila  female  14.0   
41              Ahlin, Mrs. Johan (Johanna Persdotter Larsson)  female  40.0   
42           Turpin, Mrs. William John Robert (Dorothy Ann ...  female  27.0   
44                    Laroche, Miss. Simonne Marie Anne Andree  female   3.0   
45           

### **Q3. 当时船上有多少男性持有二级船票（Pclass=2）？**
- 104
- 108
- 112
- 125

In [41]:
# 在此编写代码回答问题
#
data[(data["Sex"]=="male")&(data["Pclass"]==2)]

Survived  Pclass                                            Name  \
PassengerId                                                                     
18                  1       2                    Williams, Mr. Charles Eugene   
21                  0       2                            Fynney, Mr. Joseph J   
22                  1       2                           Beesley, Mr. Lawrence   
34                  0       2                           Wheadon, Mr. Edward H   
71                  0       2                      Jenkin, Mr. Stephen Curnow   
73                  0       2                            Hood, Mr. Ambrose Jr   
79                  1       2                   Caldwell, Master. Alden Gates   
100                 0       2                               Kantor, Mr. Sinai   
118                 0       2                 Turpin, Mr. William John Robert   
121                 0       2                     Hickman, Mr. Stanley George   
123                 0       2                            Nasser, Mr. Nicholas   
135                 0       2                  Sobey, Mr. Samuel James Hayden   
136                 0       2                              Richard, Mr. Emile   
145                 0       2                      Andrew, Mr. Edgardo Samuel   
146                 0       2                    Nicholls, Mr. Joseph Charles   
149                 0       2        Navratil, Mr. Michel ("Louis M Hoffman")   
150                 0       2               Byles, Rev. Thomas Roussel Davids   
151                 0       2                      Bateman, Rev. Robert James   
179                 0       2                              Hale, Mr. Reginald   
182                 0       2                                Pernot, Mr. Rene   
184                 1       2                       Becker, Master. Richard F   
192                 0       2                           Carbines, Mr. William   
194                 1       2                      Navratil, Master. Michel M   
214                 0       2                     Givard, Mr. Hans Kristensen   
218                 0       2                    Jacobsohn, Mr. Sidney Samuel   
220                 0       2                              Harris, Mr. Walter   
222                 0       2                            Bracken, Mr. James H   
227                 1       2                       Mellors, Mr. William John   
229                 0       2                       Fahlstrom, Mr. Arne Jonas   
233                 0       2                       Sjostedt, Mr. Ernst Adolf   
...               ...     ...                                             ...   
673                 0       2                     Mitchell, Mr. Henry Michael   
674                 1       2                           Wilhelms, Mr. Charles   
675                 0       2                      Watson, Mr. Ennis Hastings   
685                 0       2               Brown, Mr. Thomas William Solomon   
686                 0       2          Laroche, Mr. Joseph Philippe Lemercier   
696                 0       2                      Chapman, Mr. Charles Henry   
706                 0       2  Morley, Mr. Henry Samuel ("Mr Henry Marshall")   
715                 0       2                           Greenberg, Mr. Samuel   
723                 0       2                    Gillespie, Mr. William Henry   
724                 0       2                         Hodges, Mr. Henry Price   
729                 0       2                 Bryhl, Mr. Kurt Arnold Gottfrid   
733                 0       2                            Knight, Mr. Robert J   
734                 0       2                      Berriman, Mr. William John   
735                 0       2                    Troupiansky, Mr. Moses Aaron   
756                 1       2                       Hamalainen, Master. Viljo   
758                 0       2                        Bailey, Mr. Percy Andrew   
792                 0       2               

### **Q4. 所有乘客Fare的中位数和标准差是多少？（保留两位小数）**
- 中位数为 14.45, 标准差为 49.69
- 中位数为 15.1, 标准差为 12.15
- 中位数为 13.15, 标准差为 35.3
- 中位数为  17.43, 标准差为 39.1

In [103]:
# 在此编写代码回答问题
#
print(round(data["Fare"].median(),2))#中位数
print(round(data["Fare"].std(),2))#标准差

14.45
49.69


### **Q5. 幸存者的平均年龄是否高于最终死亡的乘客?**
- Yes
- No

> Tips: 可以参考pandas的`groupby`命令

In [45]:
# 在此编写代码回答问题
#

data.groupby("Survived")["Age"].mean()

Survived
0    30.626179
1    28.343690
Name: Age, dtype: float64

### **Q6.年龄小于30岁的乘客是否比年龄大于60岁的乘客有更大的存活概率？ 存活概率各为多少？**
- 年轻人为22.7%，年长者为40.6% 
- 年轻人为40.6%， 年长者为22.7%
- 年轻人为35.3%，年长者为27.4%
- 年轻人为27.4%，年长者为35.3%

> Tips: 可以查看 pandas.crosstab()函数的用法

In [76]:
# 在此编写代码回答问题
#
print(round(len(data[(data["Survived"]==1)&(data["Age"]<30)])/len(data[data["Age"]<30]),2))#年龄小于30岁的存活概率
print(round(len(data[(data["Survived"]==1)&(data["Age"]>60)])/len(data[data["Age"]>60]),2))#年龄大于60岁的存活概率

0.41
0.23


### **Q7. 女性是否比男性有更大的存活概率? 男性与女性存活人数各占所在性别的百分之多少？**
- 男性为30.2%, 女性为46.2%
- 男性为35.7%, 女性为74.2%
- 男性为21.1%, 女性为46.2%
- 男性为18.9%, 女性为74.2%

In [77]:
# 在此编写代码回答问题
#
print(round(len(data[(data["Survived"]==1)&(data["Sex"]=="male")])/len(data[data["Sex"]=="male"]),2))#男性的存活概率
print(round(len(data[(data["Survived"]==1)&(data["Sex"]=="female")])/len(data[data["Sex"]=="female"]),2))#女性的存活概率

0.19
0.74


### **Q8. 男性乘客中最受欢迎的名字是什么？**
- Charles
- Thomas
- William
- John

> Tips： 请对照给出的几个选项了解名字出现的位置

In [96]:
# 在此编写代码回答问题
#
data["Name2"]=data['Name'].apply(lambda name: name.split(',')[1].split()[1])#原来英文人名的结构是教名+自取名+姓
data[data["Sex"]=="male"]["Name2"].mode()#求众数

0    William
dtype: object

### **Q9. 男性/女性的平均年龄与`Pclass`的关系，如下声明正确的是(多选)：**
- 平均而言，一级仓中的男性年龄平均超过40岁
- 一级仓中女性的平均年龄超过40岁
- 所有不同级别仓中男性的平均年龄都大于同一级仓中的女性
- 平均而言，一级仓乘客年龄大于二级票仓，二级仓大于三级票仓

In [124]:
# 在此编写代码回答问题
#
print(data[(data["Pclass"]==1)&(data["Sex"]=="male")]["Age"].mean())#对
print(data[(data["Pclass"]==1)&(data["Sex"]=="female")]["Age"].mean())#错
print(pd.crosstab(data['Pclass'],data['Sex'],data['Age'],aggfunc="mean"))#对

41.28138613861386
34.61176470588235
Sex        female       male
Pclass                      
1       34.611765  41.281386
2       28.722973  30.740707
3       21.750000  26.507589
